akka-cluster-bootstrap/src/main/scala/com/lightbend/rp/akkaclusterbootstrap/ClusterStatusCheck.scala

```diff
@@ -23,15 +23,22 @@ import scala.concurrent.{ ExecutionContext, Future }
 
 class ClusterStatusCheck extends ReadinessCheck with HealthCheck {
   def healthy(actorSystem: ExtendedActorSystem)(implicit ec: ExecutionContext): Future[Boolean] =
-    isUp(actorSystem)
+    status(actorSystem) map {
+      case MemberStatus.Joining  => true
+      case MemberStatus.WeaklyUp => true
+      case MemberStatus.Up       => true
+      case MemberStatus.Leaving  => true
+      case MemberStatus.Exiting  => true
+      case MemberStatus.Down     => true
```

Comment:
down I guess one could say is unhealthy?
leaving and exiting let's leave as healthy since this means they still need some time to move to down...?

WDYT @patriknw

Reply:
So what happens if we tell Kubernetes that our Akka node is no longer k8s "healthy"? The documentation says:

> The kubelet uses liveness probes to know when to restart a Container. For example, liveness probes could catch a deadlock, where an application is running, but unable to make progress. Restarting a Container in such a state can help to make the application more available despite bugs.

To me this description sounds dangerously similar to the dreaded auto-downing feature. Especially if there might be false positives, it might make sense to just not use liveliness check at all. If the member status is `Down` or `Removed` the normal course of action Akka Cluster would take is to bring the application down gracefully, without someone yanking on the plug, correct?

Reply:
> To me this description sounds dangerously similar to the dreaded auto-downing feature.

It's more tricky than comparing it directly like this. 

There's two layers here, and a number of phases: Firstly, if the Akka cluster decides to mark a things as Down – it's already an internal consensus about this – this decision is reached either by a) manual downing b) auto-downing (sucks) c) split-brain-resolver (yay, good). S

> Especially if there might be false positives, it might make sense to just not use liveliness check at all.

This is impossible with regards to the Down state; once a node is declared Down it will remain such forever, it can never move back to a healthy state. So this is why marking a Down one as unhealthy is likely good, but (!), you are right about the graceful shutdown so it gets trickier:

> If the member status is Down or Removed the normal course of action Akka Cluster would take is to bring the application down gracefully, without someone yanking on the plug, correct?

Yes, we'll run coordinated shutdown and terminate the jvmhttps://doc.akka.io/docs/akka/2.5.3/scala/actors.html#coordinated-shutdown :

```
    # Run the coordinated shutdown from phase 'cluster-shutdown' when the cluster
    # is shutdown for other reasons than when leaving, e.g. when downing. This
    # will terminate the ActorSystem when the cluster extension is shutdown.
    run-coordinated-shutdown-when-down = on
```

Yet at the same time logically to me it seems useful to mark such nodes as not healthy, so you know that's why they're getting recycled. It is not k8s monitoring that decides that they have to be recycled per se, but the cluster decided so, and k8s noticed that the node marks itself as Down.

Now the actual reason to mark such node as unhealthy is about what if we get stuck in the shutdown or it takes ages for some reason? Maybe the reason the node became unresponsive is that it's completely overloading itself and shutting down will be hindered perhaps somehow and taking ages to complete as well. 

The idea: I'd like k8s to act as a safeguard here that if we don't manage to shut down gracefully in time, it should kill the box and recycle it, so we get new nodes into the cluster quickly, instead of waiting on the stuck node. Ofc this is "edge case", but all those scenarios are about when things fail, are stuck or/and unresponsive. 

So the tricky thing is to configure the latency of the healthcheck -> kill node forcibly to be larger than the timeout we give the app to complete the graceful shutdown. Can we configure this timeout in k8s?

Reply:
> So the tricky thing is to configure the latency of the healthcheck -> kill node forcibly to be larger than the timeout we give the app to complete the graceful shutdown. Can we configure this timeout in k8s?

There are a few knobs that YAML can specify, that can get timeout indirectly: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes

> - `initialDelaySeconds`: Number of seconds after the container has started before liveness or readiness probes are initiated.
> - `periodSeconds`: How often (in seconds) to perform the probe. Default to 10 seconds. Minimum value is 1.
> - `failureThreshold`: When a Pod starts and the probe fails, Kubernetes will try `failureThreshold` times before giving up. Giving up in case of liveness probe means restarting the Pod. In case of readiness probe the Pod will be marked Unready. Defaults to 3. Minimum value is 1.

So by default, the `periodSeconds` is 10s and `failureThreshold` is 3, so once "unhealthy" state is detected by k8s, we have 20s to shut the app down gracefully. We could try to increase `failureThreshold` a bit, but it might depend on the size of the cluster and what the shutdown phase looks like. I am hoping it doesn't involve saving too the data into a database.